File: 367.valid-perfect-square.cpp
/*
 * @lc app=leetcode id=367 lang=cpp
 *
 * [367] Valid Perfect Square
 */

// @lc code=start
constexpr int UPPER_BOUND = 46341; // 46341^2 > 2^31 - 1 > 46340^2
class Solution {
public:
    bool isPerfectSquare(int num) {
        int l = 0;
        int r = UPPER_BOUND;
        while (r > l) {
            int mid = l + (r - l) / 2;
            if (mid * mid == num) {
                return true;
            } else if (mid * mid < num) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return false;
    }
};
// @lc code=end

File: 367.valid-perfect-square.rs
/*
 * @lc app=leetcode id=367 lang=rust
 *
 * [367] Valid Perfect Square
 */

// @lc code=start
static UPPER_BOUND : i32 = 46341;

impl Solution {
    pub fn is_perfect_square(num: i32) -> bool {
        let mut l = 0;
        let mut r = UPPER_BOUND;
        while r > l {
            let mid = (l + (r - l) / 2).floor() as i32;
            if mid * mid == num {
                return true;
            } else if mid * mid < num {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return false;
    }
}
// @lc code=end

